allocate.go

```diff
@@ -136,7 +140,8 @@ func allocationError(hash *cid.Cid, needed, wanted int, candidatesValid []peer.I
 func (c *Cluster) obtainAllocations(
 	hash *cid.Cid,
 	rplMin, rplMax int,
-	currentValidMetrics, candidatesMetrics map[peer.ID]api.Metric) ([]peer.ID, error) {
+	currentValidMetrics, candidatesMetrics map[peer.ID]api.Metric,
```

Comment:
Would it be worth considering a metric struct that contains the current, candidate and priority metrics maps? Or possibly even better (though I haven't read through all this code, so take this proposal with a grain of salt) would be to define a custom key type, e.g. 
```go
type MetricType int

const (
    Unknown MetricType = iota
    CurrentValid
    Candidates
    Priority
)

type MetricKey struct {
    peer.ID
    Type MetricType
}
```

usage:
```go
 func (c *Cluster) obtainAllocations(hash *cid.Cid, 	rplMin, rplMax int, metrics map[MetricKey]api.Metric) ([]peer.ID, error) {...}
```

@ZenGround0 Would love to hear your thoughts on this?

Reply:
@lanzafame While we could certainly do this I'm not sure I see the advantages of these two approaches over the existing approach.  Let me know if I'm overlooking advantages to these approaches in the paragraphs below.

A struct wrapping the dictionaries would make function arguments more concise but right now I don't necessarily see this as worth the trouble of introducing a new type.

As I understand it the content of the metric struct is independent of the cluster's consideration of this metric as priority, candidate or currently valid during allocation.  The later categorization depends on the context of the call to `Cluster.allocate`, i.e. the blacklist, priority-list, pinset of the cluster, whereas the metric struct is only in charge of reporting the state of the world at a certain peer to other peers in the cluster and can, potentially anyway, be used for more than just allocation decisions.  For this reason I don't think the MetricType typed field quite belongs in the metric.

@hsanjuan let me know if I'm missing something here.